### PR TITLE
Update `DATE_ADD`/`ADDDATE` and `DATE_SUB`/`SUBDATE` functions. (#122)

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Signed-off-by: Yury-Fridlyand <yuryf@bitquilltech.com>

## Description
I referred to [MySQL docs](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html) and tried to reproduce MySQL v.8.0.30 behavior as a reference.

See team review and discussion in https://github.com/Bit-Quill/opensearch-project-sql/pull/122

### Rework on `DATE_ADD`/`ADDDATE` functions.
<details>
<summary>Changes</summary>

#### Fixes:
1. Function accepts `TIME`, considering today's date. If argument is `DATE`, it interpreted as midnight of this `DATE`.
2. Function returns always `DATETIME`.
3. Function can interact with any combination of argument types and date and time intervals both.

#### Future changes (TODOs):
1. Rework after https://github.com/opensearch-project/sql/issues/859
2. Accept and return `STRING`

#### Removed function `DATE_ADD(date, int)`
Left function (matches MySQL):
```
ADDDATE(date, interval)
DATE_ADD(date, interval)
ADDDATE(date, int)
```
#### Update signature list
1. Added `TIME`
5. Removed `STRING` - all datetime types have implicit cast from a string already

<details>
<summary>More details</summary>

<table>
<tr><td>Function</td><td>Before</td><td>After</td></tr>
<tr><td>

`DATE_ADD`
</td><td>

```
(STRING/DATE/DATETIME/TIMESTAMP, INTERVAL) -> DATETIME
```
</td><td>

```
(DATE/DATETIME/TIMESTAMP/TIME, INTERVAL) -> DATETIME
```
</td></tr><tr><td>

`ADDDATE`
</td><td>

```
(DATE, LONG) -> DATE
(STRING/DATETIME/TIMESTAMP, LONG) -> DATETIME
```
</td><td>

```
(DATE, LONG) -> DATE
(TIME/DATETIME/TIMESTAMP, LONG) -> DATETIME
```
</td></tr></table>

</details>
</details>

### Rework on `DATE_SUB`/`SUBDATE` functions.

<details>
<summary>Changes</summary>

#### The same fixes and notes as for `DATE_ADD`/`ADDDATE` functions.
</details>

### Test queries:

**NB:** Due to #853 you can't test with `DATETIME` (without code changes), but I tested with it

<details>
<summary>OpenSearch</summary>

```sql
SELECT date0, DATE_ADD(CAST(date0 AS date), INTERVAL 1 HOUR) AS `date + 1h`,
       time0, DATE_ADD(CAST(time0 AS datetime), INTERVAL 1 HOUR) AS `datetime + 1h`,
	   time1, DATE_ADD(CAST(time1 AS time), INTERVAL 1 HOUR) AS `time + 1h`,
   datetime0, DATE_ADD(CAST(datetime0 AS timestamp), INTERVAL 1 HOUR) AS `timestamp + 1h` FROM calcs LIMIT 6;
   
SELECT date0, DATE_ADD(CAST(date0 AS date), INTERVAL 1 DAY) AS `date + 1d`,
       time0, DATE_ADD(CAST(time0 AS datetime), INTERVAL 1 DAY) AS `datetime + 1d`,
	   time1, DATE_ADD(CAST(time1 AS time), INTERVAL 1 DAY) AS `time + 1d`,
   datetime0, DATE_ADD(CAST(datetime0 AS timestamp), INTERVAL 1 DAY) AS `timestamp + 1d` FROM calcs LIMIT 6;
   
SELECT date0, DATE_ADD(CAST(date0 AS date), INTERVAL 1 HOUR) AS `date + 1h`, DATE_ADD(CAST(date0 AS date), INTERVAL 1 DAY) AS `date + 1d` FROM calcs LIMIT 6;
SELECT time0, DATE_ADD(CAST(time0 AS datetime), INTERVAL 1 HOUR) AS `datetime + 1h`, DATE_ADD(CAST(time0 AS datetime), INTERVAL 1 DAY) AS `datetime + 1d` FROM calcs LIMIT 6;
SELECT time1, DATE_ADD(CAST(time1 AS time), INTERVAL 24 HOUR) AS `time + 24h` FROM calcs LIMIT 6;
SELECT datetime0, DATE_ADD(CAST(datetime0 AS timestamp), INTERVAL 1 HOUR) AS `timestamp + 1h`, DATE_ADD(CAST(datetime0 AS timestamp), INTERVAL 1 DAY) AS `timestamp + 1d` FROM calcs LIMIT 6;

SELECT time1, CAST(time1 AS time), ADDDATE(CAST(time1 AS time), 1) FROM calcs LIMIT 6;
SELECT date0, CAST(date0 AS date), ADDDATE(CAST(date0 AS date), 1) FROM calcs LIMIT 6;
SELECT time0, CAST(time0 AS datetime), ADDDATE(CAST(time0 AS datetime), 1) FROM calcs LIMIT 6;
SELECT datetime0, CAST(datetime0 AS timestamp), ADDDATE(CAST(datetime0 AS timestamp), 1) FROM calcs LIMIT 6;
```
</details>
<details>
<summary>MySQL</summary>

```sql
SELECT date0, DATE_ADD(date0, INTERVAL 1 HOUR) AS `date + 1h`, time0, DATE_ADD(time0, INTERVAL 1 HOUR) AS `datetime + 1h`, time1, DATE_ADD(time1, INTERVAL 1 HOUR) AS `time + 1h`, datetime0, DATE_ADD(datetime0, INTERVAL 1 HOUR) AS `timestamp + 1h` FROM calcs LIMIT 6;
SELECT date0, DATE_ADD(date0, INTERVAL 1 DAY) AS `date + 1d`, time0, DATE_ADD(time0, INTERVAL 1 DAY) AS `datetime + 1d`, time1, DATE_ADD(time1, INTERVAL 1 DAY) AS `time + 1d`, datetime0, DATE_ADD(datetime0, INTERVAL 1 DAY) AS `timestamp + 1d` FROM calcs LIMIT 6;

SELECT time1, DATE_ADD(time1, INTERVAL 20 HOUR) AS `time + 20h` FROM calcs LIMIT 6;
SELECT date0, DATE_ADD(date0, INTERVAL 1 HOUR) AS `date + 1h`, DATE_ADD(date0, INTERVAL 1 DAY) AS `date + 1d` FROM calcs LIMIT 6;
SELECT time0, DATE_ADD(time0, INTERVAL 1 HOUR) AS `datetime + 1h`, DATE_ADD(time0, INTERVAL 1 DAY) AS `datetime + 1d` FROM calcs LIMIT 6;
SELECT datetime0, DATE_ADD(datetime0, INTERVAL 1 HOUR) AS `timestamp + 1h`, DATE_ADD(datetime0, INTERVAL 1 DAY) AS `timestamp + 1d` FROM calcs LIMIT 6;

SELECT time1, ADDDATE(time1, 1) FROM calcs LIMIT 6;
SELECT date0, ADDDATE(date0, 1) FROM calcs LIMIT 6;
SELECT time0, ADDDATE(time0, 1) FROM calcs LIMIT 6;
SELECT datetime0, ADDDATE(datetime0, 1) FROM calcs LIMIT 6;
```
</details>

<details>
<summary>PostgreSQL</summary>

```sql
SELECT time1, pg_typeof(time1), "time + 5h", pg_typeof("time + 5h"), "time + 1d", pg_typeof("time + 1d") FROM (SELECT time1, time1 + INTERVAL '5 HOURS' AS "time + 5h", time1 + INTERVAL '1 DAYS' AS "time + 1d" FROM calcs LIMIT 6) foo;

SELECT date0, pg_typeof(date0), "date + 5h", pg_typeof("date + 5h"), "date + 1d", pg_typeof("date + 1d") FROM (SELECT date0, date0 + INTERVAL '5 HOURS' AS "date + 5h", date0 + INTERVAL '1 DAYS' AS "date + 1d" FROM calcs LIMIT 6) foo;
```
</details>

## Test data

I found that first 6 rows from `date0`, `time0`, `time1`, `datetime0` are good for testing - these columns have different data types in MySQL. In OpenSearch SQL all `[date][time]` columns have `timestamp` type, so I use `CAST` for clear testing.
<details>
<summary>data</summary>

```
mysql> show fields from Calcs where field IN ('date0', 'time0', 'time1', 'datetime0');
+-----------+-----------+------+-----+---------+-------+
| Field     | Type      | Null | Key | Default | Extra |
+-----------+-----------+------+-----+---------+-------+
| date0     | date      | YES  |     | NULL    |       |
| time0     | datetime  | YES  |     | NULL    |       |
| time1     | time      | YES  |     | NULL    |       |
| datetime0 | timestamp | YES  |     | NULL    |       |
+-----------+-----------+------+-----+---------+-------+
4 rows in set (0.00 sec)
mysql> select date0, time0, time1, datetime0 from calcs;
+------------+---------------------+----------+---------------------+
| date0      | time0               | time1    | datetime0           |
+------------+---------------------+----------+---------------------+
| 2004-04-15 | 1899-12-30 21:07:32 | 19:36:22 | 2004-07-09 10:17:35 |
| 1972-07-04 | 1900-01-01 13:48:48 | 02:05:25 | 2004-07-26 12:30:34 |
| 1975-11-12 | 1900-01-01 18:21:08 | 09:33:31 | 2004-08-02 07:59:23 |
| 2004-06-04 | 1900-01-01 18:51:48 | 22:50:16 | 2004-07-05 13:14:20 |
| 2004-06-19 | 1900-01-01 15:01:19 | NULL     | 2004-07-28 23:30:22 |
| NULL       | 1900-01-01 08:59:39 | 19:57:33 | 2004-07-22 00:30:23 |
| NULL       | 1900-01-01 07:37:48 | NULL     | 2004-07-28 06:54:50 |
| NULL       | 1900-01-01 19:45:54 | 19:48:23 | 2004-07-12 17:30:16 |
| NULL       | 1900-01-01 09:00:59 | 22:20:14 | 2004-07-04 22:49:28 |
| NULL       | 1900-01-01 20:36:00 | NULL     | 2004-07-23 21:13:37 |
| NULL       | 1900-01-01 01:31:32 | 00:05:57 | 2004-07-14 08:16:44 |
| NULL       | 1899-12-30 22:15:40 | 04:40:49 | 2004-07-25 15:22:26 |
| NULL       | 1900-01-01 13:53:46 | 04:48:07 | 2004-07-17 14:01:56 |
| NULL       | 1900-01-01 04:57:51 | NULL     | 2004-07-19 22:21:31 |
| NULL       | 1899-12-30 22:42:43 | 18:58:41 | 2004-07-31 11:57:52 |
| NULL       | 1899-12-30 22:24:08 | NULL     | 2004-07-14 07:43:00 |
| NULL       | 1900-01-01 11:58:29 | 12:33:57 | 2004-07-28 12:34:28 |
+------------+---------------------+----------+---------------------+
17 rows in set (0.00 sec)
```
</details>

## Related issue:
Fixes https://github.com/opensearch-project/sql/issues/291
https://forum.opensearch.org/t/subdate-date-sub-query-method-not-supported/9252/2

## Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/data/model/ExprTimestampValue.java [hunk 1]
```diff
@@ -30,7 +30,7 @@ public class ExprTimestampValue extends AbstractExprValue {
   /**
    * todo. only support UTC now.
    */
-  private static final ZoneId ZONE = ZoneId.of("UTC");
+  public static final ZoneId ZONE = ZoneId.of("UTC");
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/data/model/ExprTimestampValue.java:33]: just wondering should we get this info from `FunctionProperties` ?

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/data/model/ExprTimestampValue.java:33]: That means we have to deliver `FunctionProperties` everywhere `ExprTimestampValue` created or used. It is too big change to be in this PR.

Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/data/model/ExprTimestampValue.java:33]: Sure, just curious about how to handle this in future. I recall there are info in SQL session such as current catalog/schema, user/role, time zone, language etc. Probably we need to think about how to do this with FunctionProperties or a new Session abstraction. Thanks!

Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/data/model/ExprTimestampValue.java:33]: Definitely want to do this with FunctionProperties if we can.  If we end up sending tasks to multiple nodes with query details, we will want to include the execution start time (in FunctionProperties) as well as the zone id (again, in FunctionProperties).  The design of the FunctionProperties is that it is expandable to contain all the query execution details. 
cc: @MaxKsyunz

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 2]
```diff
@@ -207,30 +218,51 @@ private FunctionResolver current_date() {
...
+    );
+  }
+
+  /**
+   * A common signature for `adddate` and `subdate`.
+   * Adds/subtracts an integer number of days to/from the first argument.
+   * (DATE, LONG) -> DATE
+   * (TIME/DATETIME/TIMESTAMP, LONG) -> DATETIME
+   */
+  private Stream<SerializableFunction<?, ?>> get_adddate_subdate_signatures(
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:249]: seems cratedb normalized to arthemetic operation, https://github.com/crate/crate/blob/5.0/server/src/main/java/io/crate/expression/scalar/arithmetic/ArithmeticFunctions.java
could we do something similar?

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 3]
```diff
@@ -207,30 +218,51 @@ private FunctionResolver current_date() {
...
    */
-  private DefaultFunctionResolver add_date(FunctionName functionName) {
-    return define(functionName,
-        impl(nullMissingHandling(DateTimeFunction::exprAddDateInterval),
-            DATETIME, STRING, INTERVAL),
-        impl(nullMissingHandling(DateTimeFunction::exprAddDateInterval), DATETIME, DATE, INTERVAL),
-        impl(nullMissingHandling(DateTimeFunction::exprAddDateInterval),
+  private Stream<SerializableFunction<?, ?>> get_date_add_date_sub_signatures(
+      SerializableTriFunction<FunctionProperties, ExprValue, ExprValue, ExprValue> function) {
+    return Stream.of(
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:233]: Does Datetime required? I found you post same concern in https://github.com/opensearch-project/sql/issues/1176.

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:233]: Unfortunately, yes, unless we have 4 types for date/time.

**Comment by Yury-Fridlyand** [comment]
Rebased to resolve conflicts

**Review by penghuo — DISMISSED** [review]

**Comment by Yury-Fridlyand** [comment]
Resolved merge conflicts.

**Review by penghuo — DISMISSED** [review]

**Comment by Yury-Fridlyand** [comment]
... and again.

**Review by penghuo — APPROVED** [review]

**Review by dai-chen — APPROVED** [review]

**Review by penghuo — APPROVED** [review]

**Pull request merged**